### PR TITLE
[Feat-BugFix] Resolve custom DataLoader

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR fixes the re-instantiation of custom DataLoader and provides a better error message

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -1628,3 +1629,90 @@ def test_pytorch_profiler_nested(tmpdir):
 
         expected_c = ['add']
         assert [e.name for e in pa['c']] == expected_c
+
```
Comment (justusschock) [inline tests/trainer/test_trainer.py:1632]: can we move this to the data loading tests and not to general trainer tests?

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -1628,3 +1629,90 @@ def test_pytorch_profiler_nested(tmpdir):
...
+    model = TestModel(num_dl_idx, save_preds_on_dl_idx, failure)
+    model.test_epoch_end = None
+
+    trainer = Trainer(**trainer_args)
+    if failure == 1:
+        try:
+            _ = trainer.test(model)
+        except MisconfigurationException as e:
+            assert "Missing attributes are {'num_features'}." in str(e)
+            return
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1694]: use `with pytest.raises(...):`

File: pytorch_lightning/trainer/data_loading.py [hunk 3]
```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
         skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
+        skip_valid_keys = ['args', 'kwargs', 'self']
+
+        params = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
+        contains_dataset = True
+
+        if type(dataloader) is not DataLoader:
+            contains_dataset = "dataset" in valid_kwargs
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:120]:
```suggestion

        if "dataset" not in valid_kwargs:
            dl_args.pop('dataset')
```

We don't need to check if its a DataLoader

This change also requires removing the pop at the bottom

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
 
     def replace_sampler(self, dataloader, sampler):
         skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
+        skip_valid_keys = ['args', 'kwargs', 'self']
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:112]:
```suggestion
        skip_signature_keys = ['args', 'kwargs', 'self']
```

File: pytorch_lightning/trainer/data_loading.py [hunk 5]
```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
+
+        params = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:114]:
```suggestion
        attrs = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
```

File: pytorch_lightning/trainer/data_loading.py [hunk 6]
```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:116]:
```suggestion
        params = set(inspect.signature(dataloader.__init__).parameters)
```

File: pytorch_lightning/trainer/data_loading.py [hunk 7]
```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
 
         dl_args = {
-            k: v for k, v in dataloader.__dict__.items() if not k.startswith('_') and k not in skip_keys
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in skip_keys
         }
-
         dl_args['sampler'] = sampler
         dl_args['shuffle'] = False
+        dl_args['batch_sampler'] = None
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:129]: Does this mean that any batch_samplers get replaced? why?

File: pytorch_lightning/trainer/data_loading.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:120]: Because if it is custom DataLoader, it is possible the user don't have dataset in its __init__, and therefore we need to remove it for injection.

Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:120]: I think this change will break the code.

File: pytorch_lightning/trainer/data_loading.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py]: great catch ! I added support for it, assuming it follows the same API

File: tests/trainer/test_data_loading.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    trainer = Trainer(**trainer_args)
+    if mode < 3:
+        if mode == 1:
+            match = "Missing attributes are"
+        else:
+            match = "DistributedSampler within"
+        with pytest.raises(MisconfigurationException, match=match):
+            _ = trainer.test(model)
+    else:
+        _ = trainer.test(model)
```
Comment (carmocca) [inline tests/trainer/test_data_loading.py:110]:
```suggestion
        match = "Missing attributes are" if mode == 1 else "DistributedSampler within"
        with pytest.raises(MisconfigurationException, match=match):
            trainer.test(model)
    else:
        trainer.test(model)
```

**Review by carmocca — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
